lib/router/src/lib.rs

```diff
@@ -208,43 +247,80 @@ where
         Self::error(error::NotRecognized.into())
     }
-
-    fn no_capacity(capacity: usize) -> Self {
-        Self::error(error::NoCapacity(capacity).into())
-    }
 }
 
-impl<Req, Svc> Future for ResponseFuture<Req, Svc>
+impl<Req, Rec, Mk> Future for ResponseFuture<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req> + Clone,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    type Item = <LoadShed<Svc> as svc::Service<Req>>::Response;
+    type Item = <LoadShed<Mk::Value> as svc::Service<Req>>::Response;
     type Error = error::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
-                        Async::NotReady => {
-                            unreachable!("load shedding services must always be ready");
-                        }
-                        Async::Ready(()) => {
-                            let req = req.take().expect("polled after ready");
-                            State::Called(svc.call(req))
-                        }
+                State::Inserting(ref mut rr) => {
+                    // Aquire the lock for the router cache
+                    let mut cache = match rr.cache.poll_lock() {
+                        Async::Ready(aquired) => aquired,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    let request = rr.request.take().expect("polled after ready");
+                    let target = rr.target.take().expect("polled after ready");
+
+                    // If the target is already cached, route the request to
+                    // the service; otherwise, try to insert it
+                    if let Some(service) = cache.get(&target) {
+                        State::Calling(Some(request), Some(service))
+                    } else {
+                        // Ensure that there is capacity for a new slot
+                        cache.poll_insert()?;
+
+                        // Make a new service for the target
+                        let make = rr.make.take().expect("polled after ready");
+                        let service = LoadShed::new(make.make(&target));
+
+                        cache.insert(target, service.clone());
+                        State::Calling(Some(request), Some(service))
                     }
                 }
-                State::Called(ref mut fut) => return fut.poll().map_err(error::Error::from),
+                State::Calling(ref mut request, ref mut service) => {
+                    let mut service = service.take().expect("polled after ready");
+                    service.poll_ready()?;
```

Comment:
What if this returns NotReady?

Reply:
We know that `service` will resolve to a `LoadShed` which should always be ready. The [current implementation](https://github.com/linkerd/linkerd2-proxy/blob/35df8ab451b1dd6f7c8999e10d0e5928ddcc6231/lib/router/src/lib.rs#L231) is explicit about that and I probably should have kept that in there. I've added it back in